subsys/bluetooth/host/id.c

```diff
@@ -788,13 +793,62 @@ void bt_id_pending_keys_update(void)
 	}
 }
 
+struct bt_rl_conflict {
+	struct bt_keys *candidate;
+	struct bt_keys *found;
+};
+
+/* The Controller Resolve List is constrained by 7.8.38 "LE Add Device To Resolving List command".
+ * The Host is designed with the assumption that all local bonds can be put in the resolve list if there is room.
+ * Therefore we must refuse bonds that conflict in the resolve list.
+ * Notably, this prevents multiple local identities to bond with the same remote identity.
+ */
+void find_rl_conflict(struct bt_keys *resident, void *user_data)
+{
+	struct bt_id_conflict *conflict = user_data;
+
+	__ASSERT_NO_MSG(conflict != NULL);
+	__ASSERT_NO_MSG(conflict->candidate != NULL);
+	__ASSERT_NO_MSG(resident != NULL);
+	/* Only uncommitted bonds can be in conflict with committed bonds. */
+	__ASSERT_NO_MSG((conflict->candidate->state & BT_KEYS_ID_ADDED) == 0);
+
+	if (conflict->found) {
+		return;
+	}
+
+	/* Test against commited bonds only. */
+	if ((resident->state & BT_KEYS_ID_ADDED) == 0) {
+		return;
+	}
+
+	if(bt_addr_le_eq(&resident->addr, &conflict->candidate->addr)
+	   || bt_irk_eq(&resident->irk, &conflict->candidate->irk)) {
+		conflict->found = resident;
+	}
+}
+
+struct bt_id_conflict bt_id_find_conflict(struct bt_keys *candidate)
+{
+	struct bt_id_conflict conflict = {
+		.candidate = candidate,
+	};
+
+	bt_keys_foreach(BT_KEYS_IRK, find_rl_conflict, &conflict);
+
+	return conflict;
+}
+
 void bt_id_add(struct bt_keys *keys)
 {
 	struct bt_conn *conn;
 	int err;
 
 	BT_DBG("addr %s", bt_addr_le_str(&keys->addr));
 
+	__ASSERT_NO_MSG(keys != NULL);
+	__ASSERT_NO_MSG(!bt_id_find_conflict(keys).found);
```

Comment:
an assert for this seems pretty harsh. I would rather return an error code.

Reply:
I'm just documenting the assumption. It has always been there. Breaking this assumption was the source of this bug.

I did contemplate changing this function to "try add". But I don't think it's prettier. I suppose we would call it like in the following example?
```c
int err;
do {
    err = bt_id_add(new_bond);
    if (err == -EADDRINUSE) {
        bool allowed;
        int unpair_err;
	struct bt_keys conflict;

        BT_INFO("New bond conflict.");
	conflict = bt_id_find_conflict(new_bond);
        __ASSERT_NO_MSG(conflict);
        allowed = update_keys_check(smp, conflict.found);
        if (!allowed) {
            BT_WARN("Bond overwrite denied. Refusing new pairing.");
            return BT_SMP_ERR_AUTH_REQUIREMENTS;
        }
        BT_INFO("Bond overwrite permitted. Un-pairing old conflicting bond and finalizing new.");
        unpair_err = bt_unpair(conflict->id, &conflict->addr);
        __ASSERT_NO_MSG(!unpair_err);
    }
} while(err == -EADDRINUSE);
```